### PR TITLE
#6777 Implemented a default 30s timeout on RPC calls

### PR DESCRIPTION
**Pull request by Stono**
@Bonifacio2 My intention was to write a failing test for you to then fix, however in order to write the failing test I had to implement a timeout, and in order to implement a timeout, I ended up calling the function which tidies up queues anyway.

So what i'm saying is, I think this addresses the problem in #6777 but its worth casting your eyes over and doing some manual validation.

I'm not 100% happy with the assertion (see comments) and there is another pending test you could have a look at to if you like.

### CODE REVIEW

File: test/broker.js [hunk 1]
```diff
@@ -88,9 +110,7 @@ describe('Amqp Broker', () => {
         setTimeout(done, 500);
       });
     });
-    it('should timeout rpc requests and tidy up the replyQueue');
     it('should return an error if there are no rpc handlers for the given request');
```
Comment (Stono) [inline test/broker.js]: You could give this one a go... at the moment this scenario would just result in a timeout (as nothing is replying)

File: test/broker.js [hunk 2]
```diff
@@ -39,13 +39,35 @@ describe('Amqp Broker', () => {
...
+      });
+    });
+
+    it('should clean up rpc queue on timeout', done => {
+      let replyQueue;
+      broker.rpc({
+        filter: 'nobody.listening',
+        timeout: 100
+      }, 'message', () => {
+        should(replyQueue.consumerTag()).eql(undefined);
```
Comment (Stono) [inline test/broker.js]: All we're doing here is asserting the consumer tag is null, this is 'nulled' by the code in lib/messaging/amqp/queue.js when we stop consuming the queue however it's is solid assertion against rabbitmq that the queue is actually gone.

We could maybe look at http://www.squaremobius.net/amqp.node/channel_api.html#channel_checkQueue ?

Comment (Stono) [inline test/broker.js]: And the test should probably be in test/amqp.js actually, rather than broker.

**Comment by Stono** [comment]
I also think we should look at where .handleRpc is used and implement some sensible timeouts, for example in td.gateway we use .handleRpc when setting up a subscriber, i think a sensible timeout there for example would be 2000ms?

File: test/broker.js (hunk 2 above)
Comment (Bonifacio2) [inline test/broker.js]: There's something interesting going on here. Even if we remove the line that deletes the queue this test's still green.

File: test/broker.js (hunk 1 above)
Comment (Bonifacio2) [inline test/broker.js]: I'll give it a try and then ask for help in case i need it.
:)

**Comment by Bonifacio2** [comment]
We can talk about this manual validation you mentioned tomorrow morning.

**Comment by Stono** [comment]
@Bonifacio2 lets get this merged :p by manual testing I mean validating that the timeout works.
For example, you could do `kt scale --replicas=0 deployment/distribution` to get rid of the distributors, and then try and `curl https://gateway.integration.thoughtdata.thoughtworks.net/v1/subscription`, the command *should* timeout, and after 30s there should be no `rpc` queues in rabbitmq.

I've just gone and done it (locally):
```
$ curl http://127.0.0.1:9000/v1/subscription
{"message":"Timeout awaiting RPC Response"}%
```

And in the gateway logs:
```
  td:core:broker Timeout awaiting RPC response +30s                                                   │
  td:core:amqp:channel td.rpc.replyTo.b930d4a0-d9db-4a27-8a19-495e1aa9c304 stopping consuming queue +1 ms
```

And looking at http://127.0.0.1:15672, I don't have any rpc queues :-)

File: test/broker.js (hunk 2 above)
Comment (Bonifacio2) [inline test/broker.js]: What about this? @Stono

Comment (Stono) [inline test/broker.js]: @Bonifacio2 That's what I mean by the assertion.  We're checking that the consumerTag has been cleared, which is actually removed as part of `replyQueue.stopConsuming`, not `channel.deleteQueue(replyQueue, next)`.

We need to look at a better assertion I think, however checking a queue is present is easier said than done - this'll do for now though as it proves we're not consuming the queue (and the queues are auto deleted when there are no consumers anyway) and this is validated through the manual testing I did (see the comments on the PR)

**Pull request merged**